File: Binary_Tree_Level_Order_Traversal_II.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
 
 
class Solution {
    /**
     * @param root : The root of binary tree.
     * @return : buttom-up level order a list of lists of integer
     */
public:
    vector<vector<int>> levelOrderBottom(TreeNode *root) {
        // write your code here
        vector<vector<int>> output;
        queue<TreeNode *> queue;
        vector<int> cur_lev;
        if (root)
        {
            queue.emplace(root);
        }
        int cur_lev_cnt = queue.size();
        while (!queue.empty())
        {
            TreeNode *node = queue.front();
            --cur_lev_cnt;
            queue.pop();
            cur_lev.emplace_back(node->val);
            if (node->left)
            {
                queue.emplace(node->left);
            }
            if (node->right)
            {
                queue.emplace(node->right);
            }
            if (cur_lev_cnt == 0)
            {
                cur_lev_cnt = queue.size();
                output.emplace_back(move(cur_lev));
            }
        }
        reverse(output.begin(), output.end());
        return output;
    }
};

File: Binary Tree Level Order Traversal II.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: buttom-up level order in a list of lists of integers
    """
    def levelOrderBottom(self, root):
        # write your code here
        
        ret = []
        if not root:
            return []
        else:
            queue = [root]
            ret = []
            while queue:
                levelRet = []
                for __ in range(len(queue)):
                    node = queue.pop(0)
                    levelRet.append(node.val)
                    if node.left:
                        queue.append(node.left)
                    if node.right:
                        queue.append(node.right)
                ret.append(levelRet)
            return ret[::-1]